### DISCUSSION TITLE
Namespace Axis

### DISCUSSION THREAD

**Discussion by Devasta** [ballsteve/xrust#80] [Planning]
Hi Steve,

Just checking, do we plan to support the namespace Axis? The spec says that if you are supporting it you must make unique namespace nodes for every single element, which we can do but might impact performance. On the other hand, we'll need it for custom functions right? Or should we modify smite to run up the ancestor axis and lookup namespaces when needed?

The XSLT 3 spec says its optional and XQuery specifically excludes it.

**Comment by ballsteve** [comment]
That's a good question...

On the one hand it's a lot of extra nodes to create, with a potential impact on performance.

On the other hand it would make resolving expanded QNames much easier.

I don't think namespaces are accessed often enough to justify the cost, and running up the ancestor axis is easy and cheap.

Because of EQNames in stylesheets, I may have to rethink the strategy of transforming a stylesheet into a Transform. At the moment it throws away the stylesheet document. We may need to keep it around. Also, it may be needed for document("").

Perhaps it should be optional in smite and the parser?

**Reply by Devasta** [reply]
I'll look at adding support for namespace nodes to the parser, should be straightforward enough... I think I'll look to replace a number of the parameters on the parse() function with a single ParserConfig struct, as I am sure there'll be other features that we'll want to add later.

**Reply by Devasta** [reply]
By the way, whats the plan for intmuttree? Currently all the parser tests use Document::try_from() relating to that, want me to change that use Smite instead?

**Reply by ballsteve** [reply]
I want to retire intmuttree. smite is much improved, and I don't want to have to maintain two tree implementations.

So yes, please change parser tests to use smite.

I've made the transform, xpath, and xslt tests generic. It's a bit awkward in that you have to have the generic test, and then a test file for the concrete type that calls it. However, it means that the generic tests don't have to be changed whenever the concrete type changes.

**Comment by ballsteve** [comment]
XML Namespace support has been redesigned and reimplemented in release 1.1.
